src/main/java/java/lang/String.java

```diff
@@ -4266,8 +4266,8 @@ public static String valueOf(float f) {
      * actual program.
      */
     public static String valueOf(double d) {
-        // DIFFBLUE MODEL LIBRARY This is treated internally in CBMC
-        return CProver.nondetWithoutNullForNotModelled();
+        // DIFFBLUE MODEL LIBRARY we cast the number down to float because string solver only knows how to convert floats to string
+        return valueOf(CProver.doubleToFloat(d));
```

Comment:
:question: How is the result of this method call different from just casting `d` to type `float`?

Reply:
In `doubleToFloat` we're ensuring that there's no loss of precision in the conversion, so that we convert the correct number to string. Say we had `double x`, then converted it to `double y`, and then we converted that to `String z`. `z` is a string representation of `y`, but not necessarily of `x`. After a discussion with @romainbrenguier I went for the `doubleToFloat` implementation in order to make it more likely that the string representation of `x` equals the string representation of `y`.

Reply:
I understand the precision problem but I don't understand how `doubleToFloat` avoids this problem. As `float` has fewer bits than `double`, wouldn't there always be a loss of precision? And can we be sure that for every `double d`, there is actually a `float f` such that `d == (double) f;`? Since there are strictly fewer floats than doubles this doesn't look right... and would mean that we sometimes can't get past the `assume` in `doubleToFloat`.
Maybe we could briefly discuss this in person? It's possible that I'm just missing something.
In any case, some tests in another repository would be great to see what this actually does/fixes.